src/AWS.Deploy.Orchestration/Utilities/CloudApplicationNameGenerator.cs

```diff
@@ -0,0 +1,87 @@
+    {
+        /// <summary>
+        /// Generates a valid candidate for <see cref="CloudApplication.Name"/> based on <paramref name="target"/>.
+        /// Name  is checked to ensure it is unique, ie <paramref name="existingApplications"/> doesn't already have it.
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown if can't generate a valid name from <paramref name="target"/>.
+        /// </exception>
+        string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications);
+
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html
+        /// </summary>
+        bool IsValidName(string name);
+    }
+
+    public class CloudApplicationNameGenerator : ICloudApplicationNameGenerator
+    {
+        private readonly IFileManager _fileManager;
+        private readonly IDirectoryManager _directoryManager;
+        /// <summary>
+        /// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-using-console-create-stack-parameters.html 
+        /// </summary>
+        private readonly Regex _validatorRegex = new ("^[a-zA-Z][a-zA-Z0-9-]{2,126}$", RegexOptions.Compiled);
+
+        public CloudApplicationNameGenerator(IFileManager fileManager, IDirectoryManager directoryManager)
+        {
+            _fileManager = fileManager;
+            _directoryManager = directoryManager;
+        }
+
+        public string GenerateValidName(ProjectDefinition target, List<CloudApplication> existingApplications)
+        {
+            // generate recommendation
+            var recommendedPrefix = "deployment";
+
+            if (_fileManager.Exists(target.ProjectPath))
+                recommendedPrefix = Path.GetFileNameWithoutExtension(target.ProjectPath);
+            else if (_directoryManager.Exists(target.ProjectPath))
+                recommendedPrefix = Path.GetDirectoryName(target.ProjectPath);
+
+            // sanitize recommendation
+            recommendedPrefix =
+                new string(
+                    recommendedPrefix
+                        .ToCharArray()
+                        .SkipWhile(c => !char.IsLetter(c) && ((int)c) < 127)
+                        .Where(c =>
+                            char.IsNumber(c) ||
+                            char.IsLetter(c) && ((int)c) < 127 ||
+                            c == '-')
+                        .ToArray());
+
+            // make sure the recommendation doesn't exist already in existingApplications
+            var recommendation = recommendedPrefix;
+            var suffix = 1;
+            while (suffix < 100)
+            {
+                if (existingApplications.All(x => x.StackName != recommendation) && IsValidName(recommendation))
+                    return recommendation;
+
+                recommendation = $"{recommendedPrefix}{suffix++}";
+            }
```

Comment:
We should use something like GUID which is unique but shorter in length. Perhaps first two segments of GUID or last segment should be good enough for our use case.

Reply:
Are you recommending this?

| Current | Guid Proposal |
| --- | --- |
| Select the AWS stack: <li>WebApp</li><li>WebApp1</li><li>WebApp2</li>| Select the AWS stack: <li>WebApp</li><li>WebApp-de8bab2b</li><li>WebApp-e8da26de</li> |

I think I'd rather stick with the numbers.  I find it a bit more approachable and it has a natural sort order.

Do you want to make the case for the Guid Proposal?

Reply:
Can we check this with team just to have second opinion?

Reply:
I'm with PJ on this. This is a user facing name that the user might have to type in. Nobody wants to type in a guid. It also follows the pattern of new projects creating in VS.